plugins/synaptics-cape/fu-synaptics-cape-device.c

```diff
@@ -0,0 +1,648 @@
+						   error))) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "reset command is not supported");
+
+		return FALSE;
+	}
+
+	g_usleep(1000 * FU_SYNAPTICS_CAPE_DEVICE_USB_RESET_DELAY_MS);
+
+	g_debug("reset took %.2lfms", g_timer_elapsed(timer, NULL) * 1000);
+
+	/* success */
+	return TRUE;
+}
+
+/**
+ * fu_synaptics_cape_device_get_active_partition:
+ * @self: a #FuSynapticsCapeDevice
+ * @error: return location for an error
+ *
+ * Updates active partition information to FuSynapticsCapeDevice::active_partition
+ *
+ * Returns: retruns TRUE if operation is successful, otherwise, return FALSE if
+ *          unsuccessful.
+ *
+ **/
+static gboolean
+fu_synaptics_cape_device_setup_active_partition(FuSynapticsCapeDevice *self, GError **error)
+{
+	FuCapCmd cmd;
+
+	g_return_val_if_fail(FU_IS_SYNAPTICS_CAPE_DEVICE(self), FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.cmd_id = FU_SYNAPTICS_CMD_FW_GET_ACTIVE_PARTITION;
+	cmd.module_id = FU_SYNAPTICS_CAPE_CMD_APP_ID('C', 'T', 'R', 'L');
+
+	if (!fu_synaptics_cape_device_sendcmd_ex(self, &cmd, 0, error)) {
+		return FALSE;
+	}
+
+	self->active_partition = GUINT32_FROM_LE(cmd.data[0]);
+
+	if (self->active_partition != 1 && self->active_partition != 2) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "partition number out of range, returned partition number is %u",
+			    self->active_partition);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_cape_device_setup(FuDevice *device, GError **error)
+{
+	FuSynapticsCapeDevice *self = FU_SYNAPTICS_CAPE_DEVICE(device);
+
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_synaptics_cape_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	if (!fu_synaptics_cape_device_setup_version(self, error)) {
+		g_prefix_error(error, "failed to get firmware version info: ");
+		return FALSE;
+	}
+
+	if (!fu_synaptics_cape_device_setup_active_partition(self, error)) {
+		g_prefix_error(error, "failed to get active partition info: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_synaptics_cape_device_prepare_firmware(FuDevice *device,
+					  GBytes *fw,
+					  FwupdInstallFlags flags,
+					  GError **error)
+{
+	FuSynapticsCapeDevice *self = FU_SYNAPTICS_CAPE_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(FuFirmware) firmware = fu_synaptics_cape_firmware_new();
+	gsize offset = 0;
+
+	g_autoptr(GBytes) new_fw = NULL;
+
+	/*the "fw" includes two firmware data for each partition, we need to divide 'fw' into
+	 * two equal parts.
+	 */
+	gsize bufsz = g_bytes_get_size(fw);
+
+	if ((guint32)bufsz % 4 != 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "data not aligned to 32 bits");
+		return NULL;
+	}
+
+	/*check file size*/
```

Comment:
one space between the dots and comment please

Reply:
https://github.com/fwupd/fwupd/pull/3737

Reply:
got it.